File: 322-Coin-Change/solution.py
class Solution:
    def coinChange(self, coins, amount):
        dp = [float('inf')] * (amount + 1)

        dp[0] = 0

        for i in range(1, amount + 1):
            for j in range(len(coins)):
                if i - coins[j] >= 0:
                    dp[i] = min(dp[i], dp[i - coins[j]] + 1)

        return dp[-1] if dp[-1] != float('inf') else -1

File: 322-Coin-Change/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>

class Solution {
    public:
        int coinChange(std::vector<int> coins, int amount)
        {
            std::vector<int> dp(amount+1, amount+1);
            int n = coins.size();
            dp[0] = 0;
            for (int i = 0; i < amount + 1; i++) {
                for (int j = 0; j < n; j++) {
                    if ( coins[j] <= i) {
                        dp[i] = std::min(dp[i], dp[i - coins[j]] + 1);
                    }
                }
            }
            return dp[amount] > amount ? -1 : dp[amount];
        }
};

int main()
{
    Solution x;
    std::vector<int> a{1, 2, 5};
    int amount = 23;
    int ret = x.coinChange(a, amount);
    std::cout << ret << std::endl;
    return 1;
}
